std/math.d

```diff
@@ -5884,14 +5885,285 @@ T nextafter(T)(const T x, const T y) @safe pure nothrow @nogc
+    // The F template parameter can be used to request a specific floating-point type:
+    static assert(is(typeof(fmax!float(1u)) == float));
+    static assert(is(typeof(fmax!double(1L)) == double));
+    static assert(is(typeof(fmax!real(1.0f)) == real));
+}
 
 /****************************************
- * Returns the smaller of x and y.
+ * Returns the smallest (most negative) element of `nums`.
+ *
+ * Elements are compared after implicit conversion to a common floating-point
+ * type. If `nums` is empty or any element `isNaN`, a `nan` value will be returned.
  */
-real fmin(real x, real y) @safe pure nothrow @nogc { return x < y ? x : y; }
+auto fmin(T ...)(T nums)
+    if (allSatisfy!(ApplyRight!(isImplicitlyConvertible, real), T))
+{
+    pragma(inline, true);
+
+    alias F = AutoFImumType!T;
+    static if (T.length <= 4)
+        return fimumImpl!("<", T.length, F)(nums);
+    else
+        return fimumImpl!("<", F)(nums);
+}
+
+/// ditto
+auto fmin(F, T ...)(T nums)
+    if (isFloatingPoint!F && allSatisfy!(ApplyRight!(isImplicitlyConvertible, F), T))
+{
+    pragma(inline, true);
+
+    alias R = OriginalType!(Unqual!F);
+    static if (T.length <= 4)
+        return fimumImpl!("<", T.length, R)(nums);
+    else
+        return fimumImpl!("<", R)(nums);
+}
+
+/// ditto
+auto fmin(F)(const F[] nums) pure @safe nothrow @nogc
+    if (isFloatingPoint!F)
+{
+    pragma(inline, true);
+
+    alias R = OriginalType!(Unqual!F);
+    return fimumImpl!("<", R)(nums);
+}
+
+/// ditto
+auto fmin(F, size_t count)(auto ref const(F[count]) nums) pure @safe nothrow @nogc
+    if (isFloatingPoint!F)
+{
+    pragma(inline, true);
+
+    alias R = OriginalType!(Unqual!F);
+    static if (count <= 4)
+        return fimumImpl!("<", count, R)(nums);
+    else
+        return fimumImpl!("<", R)(nums[]);
+}
+
+///
+pure @safe nothrow @nogc unittest
+{
+    // Any number of arguments is supported:
+    assert(fmin(101.0, -5.0, 7.3723e121, double.epsilon) == -5.0);
+    assert(fmin(1) == 1.0);
+    assert(fmin(ubyte.max, 25.0f) == 25.0f);
+    // Static and dynamic floating-point arrays, too:
+    double[7] nums = [0.0, double.min_normal, 0.5, -1.0, -double.max, double.infinity, 8];
+    assert(fmin(nums) == -double.max);
+    assert(fmin(nums[1 .. 3]) == double.min_normal);
+
+    // NaN handling:
+    assert(fmin().isNaN);
+    assert(fmin(13, -431_315, long.max, real.nan, float.infinity).isNaN);
+
+    // The return type is always floating-point, and mixed signed-unsigned comparisons
+    // are safe, even if all the arguments are integers.
+    assert(fmin(-1, cast(uint)int.max) is -1.0);
+    // The F template parameter can be used to request a specific floating-point type:
+    static assert(is(typeof(fmin!float(1u)) == float));
+    static assert(is(typeof(fmin!double(1L)) == double));
+    static assert(is(typeof(fmin!real(1.0f)) == real));
+}
+
+private template AutoFImumType(T ...)
+{
```

Comment:
```D
static assert(is(CommonType!(long, float) == float));
static assert(is(AutoFImumType!(long, float) == real));
```
D allows implicit conversions from integral types to floating-point types with much less precision. Thus, I did not use `CommonType` both because:

* I want `fmax()` and `fmin()` to be reasonably precise (using at least `double` for `int` and `long`), and

* The old signature up-casts everything to `real`, which is fully precise even with `ulong` on `x86(_64)`, meaning that switching to `CommonType` would be a breaking change.